src/toxcore/tcp/connections.rs

```diff
@@ -0,0 +1,213 @@
+/*! TCP connections handling
+
+TCP connectins resides between net_crypto and tcp_client.
+It serves to net_crypto using tcp_client.
+It provides reliable connection to net_crypto via multiple tcp relays to a friend.
+When a Tox client connects to a friend via tcp relay, normally 3 redundant connections are established.
+One connection is used for data send/recv, 2 others are backup.
+In toxcore maximum number of redundant connections is 6.
+TCP connection can get into sleep mode.
+Getting into sleep mode can occur when UDP connecion is established, because Tox prefer UDP than TCP.
+When established UDP connection is disabled, TCP connecions are awaken.
+
+Defintion of struct names.
+
+<pre style="white-space:pre;">
+    `Connections` has
+        set of `ConnOfClient`(HashMap)
+        set of `Connection`(HashMap)
+    `ConnOfClient` has
+        handle of TcpClient connection
+        IP, port, PK : to save for sleeping status
+    `Connection` has
+        handles of 3 to 6 `ConnToRelay` connections for redundancy
+    `ConnToRelay` has
+        id as a key to `ConnOfClient` hashmap
+        connection_id of Routing Response packet
+</pre>
+
+*/
+
+use parking_lot::RwLock;
+
+use std::sync::Arc;
+use std::time::Instant;
+use std::collections::HashMap;
+use std::net::{IpAddr, Ipv4Addr};
+
+use toxcore::crypto_core::*;
+use toxcore::dht::server::Server;
+use toxcore::tcp::server::Client;
+use toxcore::time::*;
+
+/// The amount of maximum connections for each friend.
+pub const MAX_FRIEND_TCP_CONNECTIONS: usize =  6;
+
+/// The amount of recommended connections for each friend.
+///   NOTE: Must be at most (MAX_FRIEND_TCP_CONNECTIONS / 2).
+pub const RECOMMENDED_FRIEND_TCP_CONNECTIONS: usize =  MAX_FRIEND_TCP_CONNECTIONS / 2;
+
+/// Status of connection to TcpRelay.
+pub enum ConnectionStatus {
+    /// Uninitialized state.
+    None,
+    /// Connection is registered internally.
+    Registered,
+    /// Connection is online.
+    Online,
+}
+
+/// Status of connection to a friend.
+pub enum ConnOfClientStatus {
+    /// Uninitialized state.
+    None,
+    /// Connection is valid but not connected.
+    Valid,
+    /// Connection is valid and connected.
+    Connected,
+    /// Connection is sleeping because UDP is enabled.
+    Sleeping,
+}
+
+/// Main struct for TCP connections.
+/// Holds key_pair for TCP connections, set of TCP connections to relay,
+/// set of connections of TcpClient.
+/// PublicKey is used as a key of HashMap to avoid ABA problem.
+pub struct Connections<'a> {
+    /// Dht Server.
+    dht: &'a Server,
```

Comment:
Are you sure that you need a reference? Where are you going to store these connections? DHT server is cloneable so it might be easier to store just a `Server` object.

Reply:
What's a Dht server for?

Reply:
Now, can't find the user of dht in c-toxcore. So it is removed in struct